### ISSUE TITLE
Breaking change in Bazel CQuery due to Starlarkification of providers

### ISSUE THREAD

**Issue by guw** [bazelbuild/bazel#20473]
### Description of the bug:

The following **cquery** no longer works with Bazel 7:
```
bazel cquery "@bazel_tools//tools/jdk:current_java_toolchain" --output starlark --starlark:expr 'providers(target)["JavaToolchainInfo"].source_version'
```

It looks like that `providers(target)["JavaToolchainInfo"]` needs to be replaced with `providers(target)['@@_builtins//:common/java/java_toolchain.bzl%JavaToolchainInfo'].source_version` but it feels odd to require this verbosity now. Especially the label syntax looks subjective to further changes. It should be possible to select a provider without relying on the location.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
echo "7.0.0.rc5" > .bazelversion
```
run the query

### Have you found anything relevant by searching the web?

https://bazelbuild.slack.com/archives/CDCE4DFAP/p1702038149142419?thread_ts=1701990164.906629&cid=CDCE4DFAP

### Any other information, logs, or outputs that you want to share?

This needs to be documented. Ideally all those queries would continue to work, though.

**Comment by meisterT** [comment]
cc @gregestren

**Comment by meteorcloudy** [comment]
Maybe this is just some breaking change after we starlarkifying `JavaToolchainInfo`? @comius

**Comment by meteorcloudy** [comment]
https://github.com/bazelbuild/bazel/commit/8eecb297f965aa4ae0df732f5de4daac265f83fd /cc @hvadehra

**Comment by meteorcloudy** [comment]
Thanks for confirming!

> So we should just add this to the release notes as a breaking change.

Please do!

**Comment by gregestren** [comment]
@gu2 can you think of any clean user experience that wouldn't have the ambiguity of filtering the map? (i.e. risk of resolving to the wrong `JavaToolchainInfo` if not fully qualified)?

**Comment by guw** [comment]
It depends on how real ambiguity is. It would love to have a `find_provider(target, "JavaToolchainInfo")` helper that fails if there is none or multiple. Still trying to wrap my head around how to do this in a single line.

Perhaps there is a better way. The full query I am doing is:
```
bazel cquery --output starlark --starlark:expr "providers(target)['JavaToolchainInfo'].source_version + '::' + providers(target)['JavaToolchainInfo'].target_version + '::' + providers(target)['JavaToolchainInfo'].java_runtime.java_home" @bazel_tools//tools/jdk:current_java_toolchain
```

I would have to do the filtering three times in a single line. I don't think `--starlark:expr` allows assignments. I haven't it figured out yet.

All I want is [identify JAVA_HOME and source and target version](https://stackoverflow.com/questions/75988179/query-for-java-toolchain-details-in-bazel) so I can setup the IDE properly. Perhaps there is a better way to query this?

**Comment by hvadehra** [comment]
> I would have to do the filtering three times in a single line. I don't think --starlark:expr allows assignments. I haven't it figured out yet.

Is it necessary for the starlark code to be specified on the command line? If not, you could use `--starlark:file` instead. There are [examples](https://bazel.build/query/cquery#output-format-definition) in the docs.

**Comment by guw** [comment]
Oh, that looks doable. 

BTW, that example doc needs fixing as well:
> `providers(target)` returns a map whose keys are names of [providers](https://bazel.build/extending/rules#providers) (for example, `"DefaultInfo"`). 

It should mention the specifics around keys instability from Bazel version to version.

**Comment by guw** [comment]
Thanks @hvadehra. Using `--starlark:file` [works much better for this](https://github.com/salesforce/bazel-eclipse/commit/b017921e75582725a3083a36d1da29751b4fc850).

**Comment by hvadehra** [comment]
> we should treat @@_builtins//:common/java/java_info.bzl%JavaInfo provider name as the official way to get it in queries

Yes, for now that will the key. As noted in https://github.com/bazelbuild/bazel/issues/20473#issuecomment-1847422010 it will change (at least) once more when the java rules move out of `@builtins` and into `@rules_java` (likely in Bazel 8). Given this issue, we should keep the provider location(s) unchanged after that point in time.

**Comment by guw** [comment]
@hvadehra Is the name final now?

**Comment by hvadehra** [comment]
Yes, the java provider identities aren't expected to change with Bazel >= 8.
